README.md
# asgi-caches

[![Build Status](https://travis-ci.com/florimondmanca/asgi-caches.svg?branch=master)](https://travis-ci.com/florimondmanca/asgi-caches)
[![Coverage](https://codecov.io/gh/florimondmanca/asgi-caches/branch/master/graph/badge.svg)](https://codecov.io/gh/florimondmanca/asgi-caches)
[![Package version](https://badge.fury.io/py/asgi-caches.svg)](https://pypi.org/project/asgi-caches)

`asgi-caches` provides middleware and utilities for adding server-side HTTP caching to ASGI applications. It is powered by [`async-caches`](https://rafalp.github.io/async-caches/), and inspired by Django's cache framework.

Documentation is available at: https://asgi-caches.florimond.dev

**Note**: this project is in an "alpha" status. Several features still need to be implemented, and you should expect breaking API changes across minor versions.

## Features

- Compatibility with any ASGI application (e.g. Starlette, FastAPI, Quart, etc.).
- Support for application-wide or per-endpoint caching.
- Ability to fine-tune the cache behavior (TTL, cache control) down to the endpoint level.
- Clean and explicit API enabled by a loose coupling with `async-caches`.
- Fully type annotated.
- 100% test coverage.

## Installation

```bash
pip install "asgi-caches==0.*"
```

## Quickstart

```python
from asgi_caches.middleware import CacheMiddleware

cache = Cache("locmem://null")

async def app(scope, receive, send):
    assert scope["type"] == "http"
    headers = [(b"content-type", "text/plain")]
    await send({"type": "http.response.start", "status": 200, "headers": headers})
    await send({"type": "http.response.body", "body": b"Hello, world!"})

app = CacheMiddleware(app, cache=cache)
```

This example:

- Sets up an in-memory cache (see the [async-caches docs](https://rafalp.github.io/async-caches/) for specifics).
- Sets up an application (in this case, a raw-ASGI 'Hello, world!' app).
- Applies caching on the entire application.

To learn more, head to the [documentation](https://asgi-caches.florimond.dev).

## Credits

Due credit goes to the Django developers and maintainers, as a lot of the API and implementation was directly inspired by the [Django cache framework](https://docs.djangoproject.com/en/2.2/topics/cache/).

## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## 0.3.1 - 2019-11-23

### Changed

- Disallow applying cache middleware multiple times. (Pull #21)

## 0.3.0 - 2019-11-12

### Added

- Add `@cache_control()` decorator and its underlying middleware. (Pull #19)

## 0.2.0 - 2019-11-12

### Added

- Add `@cached()` decorator. (Pull #15)

## 0.1.1 - 2019-11-12

### Added

- Add `DEBUG` and `TRACE` logs. (Pull #14)

## 0.1.0 - 2019-11-12

### Added

- Add `CacheMiddleware`. (Pull #8)
- Prevent caching of responses that have cookies when the request has none. (Pull #9)
- Prevent caching of responses if the cache TTL is zero. (Pull #10)


tests/utils.py
import contextlib
import logging
import os
import typing

import httpx
from starlette.responses import Response
from starlette.types import ASGIApp, Message, Receive, Scope, Send

import asgi_caches.utils.logging


async def mock_receive() -> Message:
    raise NotImplementedError  # pragma: no cover


async def mock_send(message: Message) -> None:
    raise NotImplementedError  # pragma: no cover


class CacheSpy:
    def __init__(self, app: ASGIApp):
        self.app = app
        self.misses = 0

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        self.misses += 1
        await self.app(scope, receive, send)


class ComparableStarletteResponse:
    # As of 0.12, Starlette does not provide a '.__eq__()' implementation
    # for responses yet.

    def __init__(self, response: Response) -> None:
        self.response = response

    def __eq__(self, other: typing.Any) -> bool:
        assert isinstance(other, Response)
        return (
            self.response.body == other.body
            and self.response.raw_headers == other.raw_headers
            and self.response.status_code == other.status_code
        )


class ComparableHTTPXResponse:
    # As of 0.7, HTTPX does not provide a '.__eq__()' implementation
    # for responses yet.

    def __init__(self, response: httpx.models.BaseResponse) -> None:
        self.response = response

    def __eq__(self, other: typing.Any) -> bool:
        assert isinstance(other, httpx.models.BaseResponse)
        return (
            self.response.content == other.content
            and self.response.headers == other.headers
            and self.response.status_code == other.status_code
        )


@contextlib.contextmanager
def override_log_level(log_level: str) -> typing.Iterator[None]:
    os.environ["ASGI_CACHES_LOG_LEVEL"] = log_level

    # Force a reload on the logging handlers
    asgi_caches.utils.logging._logger_factory._initialized = False
    asgi_caches.utils.logging.get_logger("asgi_caches")

    try:
        yield
    finally:
        # Reset the logger so we don't have verbose output in all unit tests
        logging.getLogger("asgi_caches").handlers = []


tests/test_project.py
def test_project() -> None:
    assert True


tests/test_decorator.py
import httpx
import pytest
from caches import Cache
from starlette.applications import Starlette
from starlette.endpoints import HTTPEndpoint
from starlette.requests import Request
from starlette.responses import PlainTextResponse, Response
from starlette.types import Receive, Scope, Send

from asgi_caches.decorators import cached
from asgi_caches.middleware import CacheMiddleware
from tests.utils import CacheSpy


@pytest.mark.asyncio
async def test_decorator_raw_asgi() -> None:
    cache = Cache("locmem://null", ttl=2 * 60)

    @cached(cache)
    async def app(scope: Scope, receive: Receive, send: Send) -> None:
        response = PlainTextResponse("Hello, world!")
        await response(scope, receive, send)

    spy = app.app = CacheSpy(app.app)
    client = httpx.AsyncClient(app=app, base_url="http://testserver")

    async with cache, client:
        assert spy.misses == 0

        r = await client.get("/")
        assert r.status_code == 200
        assert r.text == "Hello, world!"
        assert "Expires" in r.headers
        assert "Cache-Control" in r.headers
        assert spy.misses == 1

        r = await client.get("/")
        assert r.status_code == 200
        assert r.text == "Hello, world!"
        assert "Expires" in r.headers
        assert "Cache-Control" in r.headers
        assert spy.misses == 1


@pytest.mark.asyncio
async def test_decorator_starlette_endpoint() -> None:
    app = Starlette()
    cache = Cache("locmem://null", ttl=2 * 60)

    @cached(cache)
    class CachedHome(HTTPEndpoint):
        async def get(self, request: Request) -> Response:
            return PlainTextResponse("Hello, world!")

    class UncachedUsers(HTTPEndpoint):
        async def get(self, request: Request) -> Response:
            return PlainTextResponse("Hello, users!")

    assert isinstance(CachedHome, CacheMiddleware)
    spy = CachedHome.app = CacheSpy(CachedHome.app)
    users_spy = CacheSpy(UncachedUsers)

    app.add_route("/", CachedHome)
    app.add_route("/users", users_spy)

    client = httpx.AsyncClient(app=app, base_url="http://testserver")

    async with cache, client:
        assert spy.misses == 0

        r = await client.get("/")
        assert r.status_code == 200
        assert r.text == "Hello, world!"
        assert "Expires" in r.headers
        assert "Cache-Control" in r.headers
        assert spy.misses == 1

        r = await client.get("/")
        assert r.status_code == 200
        assert r.text == "Hello, world!"
        assert "Expires" in r.headers
        assert "Cache-Control" in r.headers
        assert spy.misses == 1

        assert users_spy.misses == 0

        r = await client.get("/users")
        assert r.status_code == 200
        assert r.text == "Hello, users!"
        assert "Expires" not in r.headers
        assert "Cache-Control" not in r.headers
        assert users_spy.misses == 1

        r = await client.get("/users")
        assert r.status_code == 200
        assert r.text == "Hello, users!"
        assert "Expires" not in r.headers
        assert "Cache-Control" not in r.headers
        assert users_spy.misses == 2


@pytest.mark.asyncio
async def test_decorate_starlette_view() -> None:
    cache = Cache("locmem://null", ttl=2 * 60)

    with pytest.raises(ValueError):

        @cached(cache)
        async def home(request: Request) -> Response:
            ...  # pragma: no cover


src/asgi_caches/utils.py
"""
Utilities that add HTTP-specific functionality to the
otherwise protocol-agnostic features in async-caches.

The `store_in_cache()` and `get_from_cache()` helpers are the main pieces of API
defined in this module:

* `store_in_cache()` learns a cache key from a `(request, response)` pair.
* `get_from_cache()` retrieves and uses this cache key for a new `request`.
"""

import base64
import email.utils
import hashlib
import time
import typing
from urllib.request import parse_http_list

from caches import Cache
from starlette.requests import Request
from starlette.responses import Response

from .exceptions import RequestNotCachable, ResponseNotCachable

CACHABLE_METHODS = frozenset(("GET", "HEAD"))
CACHABLE_STATUS_CODES = frozenset((200, 304))
ONE_YEAR = 60 * 60 * 24 * 365


async def store_in_cache(response: Response, *, request: Request, cache: Cache) -> None:
    """
    Given a response and a request, store the response in the cache for reuse.

    To do so, a cache key is built from:

    * The absolute URL (including query parameters)
    * Varying headers [^1], as specified specified in the "Vary" header of the response.
      These headers are stored at a key that depends only on the request URL, so that
      they can be retrieved (and checked against) for future requests (without having
      to build and read an uncached response first).

    [^1]: The "Vary" header lists which headers should be taken into account in cache
    systems, because they may result in the server sending in a different response.
    For example, gzip compression requires to add "Accept-Encoding" to "Vary" because
    sending "Accept-Encoding: gzip", and "Accept-Encoding: identity" will result in
    different responses.
    """
    if response.status_code not in CACHABLE_STATUS_CODES:
        raise ResponseNotCachable(response)

    if not request.cookies and "Set-Cookie" in response.headers:
        raise ResponseNotCachable(response)

    if cache.ttl == 0:
        raise ResponseNotCachable(response)

    if cache.ttl is None:
        # From section 14.12 of RFC2616:
        # "HTTP/1.1 servers SHOULD NOT send Expires dates more than
        # one year in the future."
        max_age = ONE_YEAR
    else:
        max_age = cache.ttl

    response.headers.update(get_cache_response_headers(response, max_age=max_age))

    cache_key = await learn_cache_key(request, response, cache=cache)
    serialized_response = serialize_response(response)
    await cache.set(key=cache_key, value=serialized_response)


async def get_from_cache(
    request: Request, *, cache: Cache
) -> typing.Optional[Response]:
    """
    Given a GET or HEAD request, retrieve a cached response based on the cache key
    associated to the request.

    If no cache key is present yet, or if there is no cached response at
    that key, return `None`.

    A `None` return value indicates that the response for this
    request can (and should) be added to the cache once computed.
    """
    if request.method not in CACHABLE_METHODS:
        raise RequestNotCachable(request)

    # Try to retrieve the cached GET response (even if this is a HEAD request).
    cache_key = await get_cache_key(request, method="GET", cache=cache)
    if cache_key is None:
        return None
    serialized_response: typing.Optional[dict] = await cache.get(cache_key)

    # If not present, fallback to look for a cached HEAD response.
    if serialized_response is None:
        cache_key = await get_cache_key(request, method="HEAD", cache=cache)
        if cache_key is not None:
            serialized_response = await cache.get(cache_key)

    if serialized_response is None:
        return None

    response = deserialize_response(serialized_response)
    return response


def json_encode_bytes(data: bytes) -> str:
    """
    Given binary data, return a string representation
    that can be safely used in a JSON object.
    """
    # NOTE: we can't just return 'data.decode()', because that won't work
    # if 'data' is not in a given encoding (e.g. utf-8), as is the case
    # when using gzip compression.
    return base64.encodebytes(data).decode("ascii")


def json_decode_bytes(value: str) -> bytes:
    """
    Given a previously-computed JSON-compatible string representation of
    binary data, return the original binary data.
    """
    return base64.decodebytes(value.encode("ascii"))


def serialize_response(response: Response) -> dict:
    """Convert a response to a JSON-compatible format.

    This is required as async-caches dumps values to JSON before storing them
    in the cache system.
    """
    return {
        "content": json_encode_bytes(response.body),
        "status_code": response.status_code,
        "headers": dict(response.headers),
    }


def deserialize_response(serialized_response: dict) -> Response:
    """
    Given the JSON-compatible representation of a response, re-build the
    original response object.
    """
    return Response(
        content=json_decode_bytes(serialized_response["content"]),
        status_code=serialized_response["status_code"],
        headers=serialized_response["headers"],
    )


async def learn_cache_key(request: Request, response: Response, *, cache: Cache) -> str:
    """
    Generate a cache key from the requested absolute URL.

    Varying response headers are stored at another key based from the
    requested absolute URL.
    """
    varying_headers_cache_key = generate_varying_headers_cache_key(request, cache=cache)

    varying_headers: typing.List[str] = []
    if "Vary" in response.headers:
        for header in parse_http_list(response.headers["Vary"]):
            varying_headers.append(header.lower())
        varying_headers.sort()

    await cache.set(key=varying_headers_cache_key, value=varying_headers)

    return generate_cache_key(
        request, method=request.method, varying_headers=varying_headers, cache=cache
    )


async def get_cache_key(
    request: Request, method: str, cache: Cache
) -> typing.Optional[str]:
    """
    Given a request, return the cache key where a cached response should be looked up.
    If this request hasn't been served before, return `None` as there definitely
    won't be any matching cached response.
    """
    varying_headers_cache_key = generate_varying_headers_cache_key(request, cache=cache)
    varying_headers = await cache.get(varying_headers_cache_key)

    if varying_headers is None:
        return None

    return generate_cache_key(
        request, method=method, varying_headers=varying_headers, cache=cache,
    )


def generate_cache_key(
    request: Request, method: str, varying_headers: typing.List[str], cache: Cache,
) -> str:
    """
    Return a cache key generated from the request full URL and varying
    response headers.

    Note that the given `method` may be different from that of the request, e.g.
    because we're trying to find a response cached from a previous GET request
    while this one is a HEAD request. (This is OK because web servers will strip content
    from responses to a HEAD request before sending them on the wire.)
    """
    assert method in CACHABLE_METHODS

    ctx = hashlib.md5()
    for header in varying_headers:
        value = request.headers.get(header)
        if value is not None:
            ctx.update(value.encode())

    absolute_url = str(request.url)
    url = hashlib.md5(absolute_url.encode("ascii"))

    return cache.make_key(f"cache_page.{method}.{url.hexdigest()}.{ctx.hexdigest()}")


def generate_varying_headers_cache_key(request: Request, cache: Cache) -> str:
    """
    Return a cache key generated from the requested absolute URL, suitable for
    associated varying headers to a requested URL.
    """
    url = request.url.path
    url_hash = hashlib.md5(url.encode("ascii"))
    return cache.make_key(f"varying_headers.{url_hash.hexdigest()}")


def http_date(epoch_time: float) -> str:
    """Return a formatted date, for use in HTTP headers.

    See: https://tools.ietf.org/html/rfc7231#section-7.1.1.2
    """
    return email.utils.formatdate(epoch_time, usegmt=True)


def get_cache_response_headers(
    response: Response, *, max_age: int
) -> typing.Dict[str, str]:
    assert max_age >= 0, "Can't have a negative cache max-age"
    headers = {}

    if "Expires" not in response.headers:
        headers["Expires"] = http_date(time.time() + max_age)

    headers["Cache-Control"] = f"max-age={max_age}"

    return headers


src/asgi_caches/decorators.py
import functools
import typing

from caches import Cache
from starlette.types import ASGIApp

from .middleware import CacheControlMiddleware, CacheMiddleware
from .utils.misc import is_asgi3


def cached(cache: Cache) -> typing.Callable:
    """
    Decorator for ASGI endpoints that tries to get the response from the cache,
    or populates the cache if the response isn't cached yet.

    This decorator provides the same behavior than `CacheMiddleware`,
    but at an endpoint level.

    Raises 'ValueError' if the wrapped callable isn't an ASGI application.
    """

    def wrap(app: ASGIApp) -> ASGIApp:
        _validate_asgi3(app)
        middleware = CacheMiddleware(app, cache=cache)
        return _wrap_in_middleware(app, middleware)

    return wrap


def cache_control(**kwargs: typing.Any) -> typing.Callable:
    """
    Decorator for ASGI endpoints that patches Cache-Control directives on the response.
    """

    def wrap(app: ASGIApp) -> ASGIApp:
        _validate_asgi3(app)
        middleware = CacheControlMiddleware(app, **kwargs)
        return _wrap_in_middleware(app, middleware)

    return wrap


def _wrap_in_middleware(app: ASGIApp, middleware: ASGIApp) -> ASGIApp:
    # Use `updated=()` to prevent copying `__dict__` onto `middleware`.
    # (If `app` is a middleware itself and has a `.app` attribute, it would be copied
    # onto `middleware`, effectively removing `app` from the middleware chain.)
    return functools.wraps(app, updated=())(middleware)


def _validate_asgi3(app: ASGIApp) -> None:
    if not is_asgi3(app):
        raise ValueError(
            f"{app!r} does not seem to be an ASGI3 callable. "
            "Did you try to apply this decorator to a framework-specific view "
            "function? (It can only be applied to ASGI callables.)"
        )


src/asgi_caches/exceptions.py
from caches import Cache
from starlette.requests import Request
from starlette.responses import Response


class ASGICachesException(Exception):
    pass


class RequestNotCachable(ASGICachesException):
    """Raised when a request cannot be cached."""

    def __init__(self, request: Request) -> None:
        super().__init__()
        self.request = request


class ResponseNotCachable(ASGICachesException):
    """Raised when a response cannot be cached."""

    def __init__(self, response: Response) -> None:
        super().__init__()
        self.response = response


class DuplicateCaching(ASGICachesException):
    """
    Raised when more than one cache middleware
    were detected in the middleware stack.
    """


class CacheNotConnected(ASGICachesException):
    """
    Raised when trying to use the cache, but it isn't connected.
    """

    def __init__(self, cache: Cache) -> None:
        super().__init__(
            f"Cache at '{cache.url}' is not connected.\n"
            "HINT: https://rafalp.github.io/async-caches/backends/#connection"
        )
        self.cache = cache


src/asgi_caches/utils/misc.py
"""Miscellaneous utilities and helper functions."""

import base64
import email.utils
import inspect
import typing


def http_date(epoch_time: float) -> str:
    """Return a formatted date, for use in HTTP headers.

    See: https://tools.ietf.org/html/rfc7231#section-7.1.1.2
    """
    return email.utils.formatdate(epoch_time, usegmt=True)


def bytes_to_json_string(data: bytes) -> str:
    """
    Given binary data, return a string representation
    that can be safely used in a JSON object.
    """
    # NOTE: we can't just return 'data.decode()', because that won't work
    # if 'data' is not in a given encoding (e.g. utf-8), as is the case
    # when e.g. 'data' is gzip-compressed.
    return base64.encodebytes(data).decode("ascii")


def json_string_to_bytes(value: str) -> bytes:
    """
    Given a previously-computed JSON-compatible string representation of
    binary data, return the original binary data.
    """
    return base64.decodebytes(value.encode("ascii"))


def has_asgi3_signature(func: typing.Callable) -> bool:
    sig = inspect.signature(func)
    own_parameters = {name for name in sig.parameters if name != "self"}
    return own_parameters == {"scope", "receive", "send"}


def is_asgi3(app: typing.Any) -> bool:
    """Return whether 'app' corresponds to an ASGI3 callable."""
    if inspect.isclass(app):
        constructor = app.__init__  # type: ignore
        return has_asgi3_signature(constructor) and hasattr(app, "__await__")

    if inspect.isfunction(app):
        return inspect.iscoroutinefunction(app) and has_asgi3_signature(app)

    try:
        call = app.__call__
    except AttributeError:
        return False
    else:
        return inspect.iscoroutinefunction(call) and has_asgi3_signature(call)


def kvformat(**kwargs: typing.Any) -> str:
    return " ".join(f"{key}={value}" for key, value in kwargs.items())


src/asgi_caches/utils/logging.py
import logging
import os
import sys
import typing

try:  # pragma: no cover
    import click

    # Extra log info for optional coloured terminal outputs.
    HIT_EXTRA = {"color_message": "cache_lookup " + click.style("%s", fg="green")}
    MISS_EXTRA = {"color_message": "cache_lookup " + click.style("%s", fg="yellow")}
except ImportError:  # pragma: no cover
    HIT_EXTRA = {}
    MISS_EXTRA = {}


TRACE_LOG_LEVEL = 5


class Logger(logging.Logger):
    # Stub for type checkers.
    def trace(self, message: str, *args: typing.Any, **kwargs: typing.Any) -> None:
        ...  # pragma: no cover


class LoggerFactory:
    log_level_env_var = "ASGI_CACHES_LOG_LEVEL"
    log_line_format = "%(levelname)s [%(asctime)s] %(name)s - %(message)s"
    date_format = "%Y-%m-%d %H:%M:%S"

    def _configure_package_logger(self, log_level: str) -> None:
        assert log_level in ("DEBUG", "TRACE")
        logger = logging.getLogger("asgi_caches")
        logger.setLevel(logging.DEBUG if log_level == "DEBUG" else TRACE_LOG_LEVEL)
        handler = logging.StreamHandler(sys.stderr)
        handler.setFormatter(
            logging.Formatter(fmt=self.log_line_format, datefmt=self.date_format)
        )
        logger.addHandler(handler)

    def get(self, name: str) -> Logger:
        """
        Get a logger instance, and optionally set up logging.
        """
        if not getattr(self, "_initialized", False):
            logging.addLevelName(TRACE_LOG_LEVEL, "TRACE")
            log_level = os.environ.get(self.log_level_env_var, "").upper()
            if log_level in ("DEBUG", "TRACE"):
                self._configure_package_logger(log_level)
            self._initialized = True

        logger = logging.getLogger(name)

        def trace(message: str, *args: typing.Any, **kwargs: typing.Any) -> None:
            logger.log(TRACE_LOG_LEVEL, message, *args, **kwargs)

        logger.trace = trace  # type: ignore

        return typing.cast(Logger, logger)


_logger_factory = LoggerFactory()
get_logger = _logger_factory.get


src/asgi_caches/utils/cache.py
"""
Utilities that add HTTP-specific functionality to the
otherwise protocol-agnostic features in async-caches.

The `store_in_cache()` and `get_from_cache()` helpers are the main pieces of API
defined in this module:

* `store_in_cache()` learns a cache key from a `(request, response)` pair.
* `get_from_cache()` retrieves and uses this cache key for a new `request`.
"""

import hashlib
import time
import typing
from urllib.request import parse_http_list

from caches import Cache
from starlette.datastructures import MutableHeaders
from starlette.requests import Request
from starlette.responses import Response

from ..exceptions import RequestNotCachable, ResponseNotCachable
from .logging import get_logger
from .misc import bytes_to_json_string, http_date, json_string_to_bytes

logger = get_logger(__name__)


CACHABLE_METHODS = frozenset(("GET", "HEAD"))
CACHABLE_STATUS_CODES = frozenset((200, 304))
ONE_YEAR = 60 * 60 * 24 * 365


async def store_in_cache(response: Response, *, request: Request, cache: Cache) -> None:
    """
    Given a response and a request, store the response in the cache for reuse.

    To do so, a cache key is built from:

    * The absolute URL (including query parameters)
    * Varying headers [^1], as specified specified in the "Vary" header of the response.
      These headers are stored at a key that depends only on the request URL, so that
      they can be retrieved (and checked against) for future requests (without having
      to build and read an uncached response first).

    [^1]: The "Vary" header lists which headers should be taken into account in cache
    systems, because they may result in the server sending in a different response.
    For example, gzip compression requires to add "Accept-Encoding" to "Vary" because
    sending "Accept-Encoding: gzip", and "Accept-Encoding: identity" will result in
    different responses.
    """
    if response.status_code not in CACHABLE_STATUS_CODES:
        logger.trace("response_not_cachable reason=status_code")
        raise ResponseNotCachable(response)

    if not request.cookies and "Set-Cookie" in response.headers:
        logger.trace("response_not_cachable reason=cookies_for_cookieless_request")
        raise ResponseNotCachable(response)

    if cache.ttl == 0:
        logger.trace("response_not_cachable reason=zero_ttl")
        raise ResponseNotCachable(response)

    if cache.ttl is None:
        # From section 14.12 of RFC2616:
        # "HTTP/1.1 servers SHOULD NOT send Expires dates more than
        # one year in the future."
        max_age = ONE_YEAR
        logger.trace(f"max_out_ttl value={max_age!r}")
    else:
        max_age = cache.ttl

    logger.debug(f"store_in_cache max_age={max_age!r}")

    cache_headers = get_cache_response_headers(response, max_age=max_age)
    logger.trace(f"patch_response_headers headers={cache_headers!r}")
    response.headers.update(cache_headers)

    cache_key = await learn_cache_key(request, response, cache=cache)
    logger.trace(f"learnt_cache_key cache_key={cache_key!r}")
    serialized_response = serialize_response(response)
    logger.trace(
        f"store_response_in_cache key={cache_key!r} value={serialized_response!r}"
    )
    await cache.set(key=cache_key, value=serialized_response)


async def get_from_cache(
    request: Request, *, cache: Cache
) -> typing.Optional[Response]:
    """
    Given a GET or HEAD request, retrieve a cached response based on the cache key
    associated to the request.

    If no cache key is present yet, or if there is no cached response at
    that key, return `None`.

    A `None` return value indicates that the response for this
    request can (and should) be added to the cache once computed.
    """
    logger.trace(
        f"get_from_cache "
        f"request.url={str(request.url)!r} "
        f"request.method={request.method!r}"
    )
    if request.method not in CACHABLE_METHODS:
        logger.trace("request_not_cachable reason=method")
        raise RequestNotCachable(request)

    logger.trace("lookup_cached_response method='GET'")
    # Try to retrieve the cached GET response (even if this is a HEAD request).
    cache_key = await get_cache_key(request, method="GET", cache=cache)
    if cache_key is None:
        logger.trace("cache_key found=False")
        return None
    logger.trace(f"cache_key found=True cache_key={cache_key!r}")
    serialized_response: typing.Optional[dict] = await cache.get(cache_key)

    # If not present, fallback to look for a cached HEAD response.
    if serialized_response is None:
        logger.trace("lookup_cached_response method='HEAD'")
        cache_key = await get_cache_key(request, method="HEAD", cache=cache)
        assert cache_key is not None
        logger.trace(f"cache_key found=True cache_key={cache_key!r}")
        serialized_response = await cache.get(cache_key)

    if serialized_response is None:
        logger.trace("cached_response found=False")
        return None

    logger.trace(
        f"cached_response found=True key={cache_key!r} value={serialized_response!r}"
    )
    response = deserialize_response(serialized_response)
    return response


def serialize_response(response: Response) -> dict:
    """Convert a response to JSON format.

    (This is required as `async-caches` dumps values to JSON before storing them
    in the cache system.)
    """
    return {
        "content": bytes_to_json_string(response.body),
        "status_code": response.status_code,
        "headers": dict(response.headers),
    }


def deserialize_response(serialized_response: dict) -> Response:
    """
    Given the JSON representation of a response, re-build the
    original response object.
    """
    return Response(
        content=json_string_to_bytes(serialized_response["content"]),
        status_code=serialized_response["status_code"],
        headers=serialized_response["headers"],
    )


async def learn_cache_key(request: Request, response: Response, *, cache: Cache) -> str:
    """
    Generate a cache key from the requested absolute URL.

    Varying response headers are stored at another key based from the
    requested absolute URL.
    """
    logger.trace(
        "learn_cache_key "
        f"request.method={request.method!r} "
        f"response.headers.Vary={response.headers.get('Vary')!r}"
    )
    varying_headers_cache_key = generate_varying_headers_cache_key(request, cache=cache)

    varying_headers: typing.List[str] = []
    if "Vary" in response.headers:
        for header in parse_http_list(response.headers["Vary"]):
            varying_headers.append(header.lower())
        varying_headers.sort()

    logger.trace(
        "store_varying_headers "
        f"cache_key={varying_headers_cache_key!r} headers={varying_headers!r}"
    )
    await cache.set(key=varying_headers_cache_key, value=varying_headers)

    return generate_cache_key(
        request, method=request.method, varying_headers=varying_headers, cache=cache
    )


async def get_cache_key(
    request: Request, method: str, cache: Cache
) -> typing.Optional[str]:
    """
    Given a request, return the cache key where a cached response should be looked up.
    If this request hasn't been served before, return `None` as there definitely
    won't be any matching cached response.
    """
    logger.trace(f"get_cache_key request.url={str(request.url)!r} method={method!r}")
    varying_headers_cache_key = generate_varying_headers_cache_key(request, cache=cache)
    varying_headers = await cache.get(varying_headers_cache_key)

    if varying_headers is None:
        logger.trace("varying_headers found=False")
        return None
    logger.trace(f"varying_headers found=True headers={varying_headers!r}")

    return generate_cache_key(
        request, method=method, varying_headers=varying_headers, cache=cache,
    )


def generate_cache_key(
    request: Request, method: str, varying_headers: typing.List[str], cache: Cache,
) -> str:
    """
    Return a cache key generated from the request full URL and varying
    response headers.

    Note that the given `method` may be different from that of the request, e.g.
    because we're trying to find a response cached from a previous GET request
    while this one is a HEAD request. (This is OK because web servers will strip content
    from responses to a HEAD request before sending them on the wire.)
    """
    assert method in CACHABLE_METHODS

    ctx = hashlib.md5()
    for header in varying_headers:
        value = request.headers.get(header)
        if value is not None:
            ctx.update(value.encode())

    absolute_url = str(request.url)
    url = hashlib.md5(absolute_url.encode("ascii"))

    return cache.make_key(f"cache_page.{method}.{url.hexdigest()}.{ctx.hexdigest()}")


def generate_varying_headers_cache_key(request: Request, cache: Cache) -> str:
    """
    Return a cache key generated from the requested absolute URL, suitable for
    associating varying headers to a requested URL.
    """
    url = request.url.path
    url_hash = hashlib.md5(url.encode("ascii"))
    return cache.make_key(f"varying_headers.{url_hash.hexdigest()}")


def get_cache_response_headers(
    response: Response, *, max_age: int
) -> typing.Dict[str, str]:
    """Return caching-related headers to add to a response."""
    assert max_age >= 0, "Can't have a negative cache max-age"
    headers = {}

    if "Expires" not in response.headers:
        headers["Expires"] = http_date(time.time() + max_age)

    patch_cache_control(response.headers, max_age=max_age)

    return headers


def patch_cache_control(headers: MutableHeaders, **kwargs: typing.Any) -> None:
    """
    Patch headers with an extended version of the initial Cache-Control header by adding
    all keyword arguments to it.
    """
    cache_control: typing.Dict[str, typing.Any] = {}
    for field in parse_http_list(headers.get("Cache-Control", "")):
        try:
            key, value = field.split("=")
        except ValueError:
            cache_control[field] = True
        else:
            cache_control[key] = value

    if "max-age" in cache_control and "max_age" in kwargs:
        kwargs["max_age"] = min(int(cache_control["max-age"]), kwargs["max_age"])

    if "public" in kwargs:
        raise NotImplementedError(
            "The 'public' cache control directive isn't supported yet."
        )

    if "private" in kwargs:
        raise NotImplementedError(
            "The 'private' cache control directive isn't supported yet."
        )

    for key, value in kwargs.items():
        key = key.replace("_", "-")
        cache_control[key] = value

    directives: typing.List[str] = []
    for key, value in cache_control.items():
        if value is False:
            continue
        if value is True:
            directives.append(key)
        else:
            directives.append(f"{key}={value}")

    patched_cache_control = ", ".join(directives)

    if patched_cache_control:
        headers["Cache-Control"] = patched_cache_control
    else:
        del headers["Cache-Control"]
